### PR TITLE
[Product] Fix duplicate entry exception

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | maybe |
| Related tickets | #3570, #7446  |
| License         | MIT |

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php [hunk 1]
```diff
@@ -66,7 +69,15 @@ public function configureOptions(OptionsResolver $resolver)
             'resource' => 'sylius.taxon',
             'choice_name' => 'name',
             'choice_value' => 'code',
+            'resource_repository' => function (ServiceRegistryInterface $repositoryRegistry, Options $options) {
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php]: Second argument is not used.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php [hunk 2]
```diff
@@ -45,14 +48,14 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         if ($options['multiple']) {
             $builder->addModelTransformer(
                 new RecursiveTransformer(
-                    new ProductTaxonToTaxonTransformer($this->productTaxonFactory)
+                    new ProductTaxonToTaxonTransformer($this->productTaxonFactory, $options['get_resource_repository'], $options['product'])
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php]: Can't see `get_resource_repository` option defined anywhere.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php [hunk 3]
```diff
@@ -45,14 +48,14 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
                 )
             );
         }
 
         if (!$options['multiple']) {
             $builder->addModelTransformer(
-                new ProductTaxonToTaxonTransformer($this->productTaxonFactory)
+                new ProductTaxonToTaxonTransformer($this->productTaxonFactory, $options['get_resource_repository'], $options['product'])
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php]: Can't see `get_resource_repository` option defined anywhere.

File: src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        $taxonRepository = $this->prophesize(TaxonRepositoryInterface::class);
+        $productTaxonRepository=  $this->prophesize(RepositoryInterface::class);
+
+        $this->resourceRepositoryRegistry->get('sylius.taxon')->willReturn($taxonRepository);
+        $this->resourceRepositoryRegistry->get('sylius.product_taxon')->willReturn($productTaxonRepository);
+        $taxonRepository->findOneBy(['code' => 'mug'])->willReturn($taxon);
+        $taxonRepository->findOneBy(['code' => 'book'])->willReturn($taxon);
+        $productTaxonRepository->findOneBy(['product' => $product, 'taxon' => $taxon])->willReturn(null);
+
+
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php]: Doubled blank line.

File: src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        $taxonRepository = $this->prophesize(TaxonRepositoryInterface::class);
+        $productTaxonRepository=  $this->prophesize(RepositoryInterface::class);
+
+        $this->resourceRepositoryRegistry->get('sylius.taxon')->willReturn($taxonRepository);
+        $this->resourceRepositoryRegistry->get('sylius.product_taxon')->willReturn($productTaxonRepository);
+        $taxonRepository->findOneBy(['code' => 'mug'])->willReturn($taxon);
+        $productTaxonRepository->findOneBy(['product' => $product, 'taxon' => $taxon])->willReturn(null);
+
+
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php]: Doubled blank line.

File: src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            new PreloadedExtension([$productTaxonAutoCompleteType, $resourceAutoCompleteType], []),
+        ];
+    }
+
+    /**
+     * @test
+     */
+    public function it_creates_new_product_taxons_based_on_given_product_and_passed_taxon_codes()
+    {
+        $formData = 'mug,book';
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php]: It's used only once, let's inline it.

File: src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+        $taxon = $this->prophesize(TaxonInterface::class);
+        $product = $this->prophesize(ProductInterface::class);
+
+        $taxonRepository = $this->prophesize(TaxonRepositoryInterface::class);
+        $productTaxonRepository=  $this->prophesize(RepositoryInterface::class);
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php]: CS

File: src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+        $this->resourceRepositoryRegistry->get('sylius.taxon')->willReturn($taxonRepository);
+        $this->resourceRepositoryRegistry->get('sylius.product_taxon')->willReturn($productTaxonRepository);
+        $taxonRepository->findOneBy(['code' => 'mug'])->willReturn($taxon);
+        $taxonRepository->findOneBy(['code' => 'book'])->willReturn($taxon);
+        $productTaxonRepository->findOneBy(['product' => $product, 'taxon' => $taxon])->willReturn(null);
+
+
+        $productTaxon = $this->prophesize(ProductTaxonInterface::class);
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php]: Can we move it up to the other mocks creation?

File: src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+        $this->productTaxonFactory->createNew()->willReturn($productTaxon);
+
+        $form = $this->factory->create(ProductTaxonAutocompleteChoiceType::class, new ArrayCollection([]), [
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Form/Type/Taxon/ProductTaxonAutocompleteChoiceTypeTest.php]: `new ArrayCollection([])` -> `new ArrayCollection()`

File: src/Sylius/Bundle/ResourceBundle/Form/Type/ResourceAutocompleteChoiceType.php [hunk 10]
```diff
@@ -94,6 +95,15 @@ public function configureOptions(OptionsResolver $resolver)
...
                 'placeholder' => '',
+                'repository' => function (Options $options) {
+                    return $this->resourceRepositoryRegistry->get($options['resource']);
+                },
+                'get_resource_repository' => function (Options $options) {
+                    return $options['resource_repository']($this->resourceRepositoryRegistry, $options);
+                },
+                'resource_repository' => function (ServiceRegistryInterface $repositoryRegistry, Options $options) {
+                    return $repositoryRegistry->get($options['resource']);
+                },
```
Comment (pamil) [inline src/Sylius/Bundle/ResourceBundle/Form/Type/ResourceAutocompleteChoiceType.php]: What is the point of these three repositories-related options? WTF alert 🌮

File: src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php [hunk 11]
```diff
@@ -42,17 +45,27 @@ public function __construct(FactoryInterface $productTaxonFactory)
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $productTaxonRepository = $options['repository_registry']()->get('sylius.product_taxon');
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Taxon/ProductTaxonAutocompleteChoiceType.php]: Let's just inject `ProductTaxonRepository` here without these hacks.

File: src/Sylius/Bundle/ResourceBundle/Form/Type/ResourceAutocompleteChoiceType.php [hunk 12]
```diff
@@ -94,6 +95,12 @@ public function configureOptions(OptionsResolver $resolver)
             ->setDefaults([
                 'multiple' => false,
                 'placeholder' => '',
+                'repository' => function (Options $options) {
+                    return $this->resourceRepositoryRegistry->get($options['resource']);
+                },
+                'repository_registry' => function () {
```
Comment (pamil) [inline src/Sylius/Bundle/ResourceBundle/Form/Type/ResourceAutocompleteChoiceType.php]: And then let's remove this options.

**Comment by pjedrzejewski** [comment]
Thanks Arek, great work!

**Pull request merged**